src/soap/handlers/GetVariables.js

```diff
@@ -0,0 +1,168 @@
+  request,
+  { date = new Date(), helpers, method, uniqueid }
+) {
+  const org =
+    typeof request.params.org === 'string'
+      ? helpers.org(request.params.org)
+      : undefined
+
+  // Fetch organization
+  const organization = org
+    ? await helpers.findOneCached('organizations', '', {
+        slug: helpers.safeName(org)
+      })
+    : undefined
+
+  const unitCV = await helpers.getUnitCV()
+
+  yield soapEnvelopeStart() +
+    soapHeaderStart() +
+    soapWsaAction('GetVariablesResponse') +
+    soapWsaMessageID(uniqueid || uuid()) +
+    soapWsaRelatesTo(uniqueid || uuid()) +
+    soapWsaTo() +
+    soapWsseSecurityStart() +
+    soapWsuTimestampStart(uniqueid || uuid()) +
+    soapWsuInfo({ date }) +
+    soapWsuTimestampEnd() +
+    soapWsseSecurityEnd() +
+    soapHeaderEnd() +
+    soapBodyStart() +
+    responseStart('GetVariablesResponse') +
+    variablesResultStart() +
+    encodeXML(
+      variablesResponseStart({ isObject: false }) +
+        queryInfoStart() +
+        queryInfoType({
+          date,
+          method,
+          parameters: [['variable']]
+        }) +
+        queryInfoNote({ note: 'OD Web Service' }) +
+        queryInfoEnd()
+    )
+
+  // Fetch datastreams
+  const datastreamsParams = Object.assign(
+    {
+      is_enabled: true,
+      is_hidden: false,
+      $limit: 2000,
+      $sort: { _id: 1 }
+    },
+    organization &&
+      organization.data &&
+      organization.data.length &&
+      organization.data[0]._id
+      ? { organization_id: organization.data[0]._id }
+      : undefined
+  )
+  let datastreams = await helpers.findMany('datastreams', datastreamsParams)
+
+  if (datastreams && datastreams.length) {
+    yield encodeXML(variablesStart())
+  }
+
+  const variableCodes = new Set()
+
+  while (datastreams.length) {
+    let i = 0
+
+    for (const datastream of datastreams) {
+      const refsMap =
+        datastream && datastream.external_refs
+          ? helpers.externalRefsMap(datastream.external_refs)
+          : undefined
+      const variableCode =
+        refsMap && refsMap.get('his.odm.variables.VariableCode')
+
+      // Normalize datastreams to unique variables
+      if (variableCode && !variableCodes.has(variableCode)) {
+        variableCodes.add(variableCode)
+
+        yield encodeXML(
+          variableStart() +
+            variableInfoType({ datastream, refsMap, unitCV }) +
+            variableEnd()
+        )
+      }
+
+      // Stay async friendly; scan 200 at a time (hardcoded)
+      i++
+      if (!(i % 200)) await new Promise(resolve => setImmediate(resolve))
+    }
+
+    // Fetch next page
+    datastreams = await helpers.findMany(
+      'datastreams',
+      Object.assign(
+        {
+          _id: { $gt: datastreams[datastreams.length - 1]._id }
+        },
+        datastreamsParams
+      )
+    )
+  }
+
+  // datastream.length condition is not possible; if datastream, variableCodes has size
+  if (variableCodes.size) {
```

Comment:
There is an edge case where this can fail and you would get a variables start element and no end element. Also a bit confusing. I would just extend the "if" block from the line `if (datastreams && datastreams.length) {`. This makes it foolproof and easy to understand.

Reply:
I think a flag just complicates it. Can it just be put under one if statement block?